src/destreamer.ts

```diff
@@ -208,63 +184,84 @@ async function downloadVideo(videoUrls: string[], outputDirectory: string, sessi
 
         // Very experimental inline thumbnail rendering
         if (!argv.noThumbnails)
             await drawThumbnail(video.posterImage, session.AccessToken);
 
         console.info('Spawning ffmpeg with access token and HLS URL. This may take a few seconds...\n');
 
-        ffmpeg()
-            .input(video.playbackUrl)
-            .inputOption([
-                // Never remove those "useless" escapes or ffmpeg will not
-                // pick up the header correctly
-                // eslint-disable-next-line no-useless-escape
-                '-headers', `Authorization:\ Bearer\ ${session.AccessToken}`,
-            ])
-            .format('mp4')
-            .saveToFile(outputPath)
-            .on('codecData', data => {
-                console.log(`Input is ${data.video} with ${data.audio} audio.\n`);
-
-                pbar.start(video.duration, 0, {
-                    speed: '0'
-                });
-
-                process.on('SIGINT', () => {
-                    pbar.stop();
-                });
-            })
-            .on('progress', progress => {
-                const currentChunks = ffmpegTimemarkToChunk(progress.timemark);
-
-                pbar.update(currentChunks, {
-                    speed: progress.currentKbps
-                });
-            })
-            .on('error', err => {
-                pbar.stop();
-                console.log(`ffmpeg returned an error: ${err.message}`);
-            })
-            .on('end', () => {
+        const outputPath = outputDirectory + path.sep + video.title + '.mp4';
+        const ffmpegInpt = new FFmpegInput(video.playbackUrl, new Map([
+            ['headers', `Authorization:\ Bearer\ ${session.AccessToken}`]
+        ]));
+        const ffmpegOutput = new FFmpegOutput(outputPath);
+        const ffmpegCmd = new FFmpegCommand();
+
+        pbar.start(video.duration, 0, {
+            speed: '0'
+        });
+
+        // prepare ffmpeg command line
+        ffmpegCmd.addInput(ffmpegInpt);
+        ffmpegCmd.addOutput(ffmpegOutput);
+
+        // set events
+        ffmpegCmd.on('update', (data: any) => {
+            const currentChunks = ffmpegTimemarkToChunk(data.out_time);
+            const incChunk = currentChunks - previousChunks;
+
+            pbar.increment(incChunk, {
+                speed: data.bitrate
+            });
+
+            previousChunks = currentChunks;
+        });
+
+        ffmpegCmd.on('error', (error: any) => {
+            pbar.stop();
+            console.log(`\nffmpeg returned an error: ${error.message}`);
+            process.exit(34);
+        });
+
+        process.on('SIGINT', () => {
+            pbar.stop();
+        });
+
+        // let the magic begin...
+        await new Promise((resolve: any, reject: any) => {
+            ffmpegCmd.on('success', (data:any) => {
+                pbar.update(video.duration); // overflow, just in case we don't reach 100%
```

Comment:
It's not clear how this works, could you please comment further? What should `video.duration` return and what does `pbar.update()` expect so the progress bar completes execution?

Reply:
video.duration will be renamed to totalChunks

update() sets a new value for the progress bar
here i m setting totalChunks so progress = totalChunks, which is 100%

that comment is wrong, forgot that